util/file_util.py

```diff
@@ -0,0 +1,922 @@
+# Copyright 2023 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the 'License');
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#         https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an 'AS IS' BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Utility classes and functions for file operations.
+Supports local files, files on Google Cloud Storage (GCS) and Google Spreadsheets.
+"""
+
+import ast
+import csv
+import fnmatch
+import glob
+import gspread
+import os
+import pickle
+import pprint
+import sys
+import tempfile
+
+from absl import logging
+from google.cloud import storage
+from typing import Union
+
+
+class FileIO:
+    '''Class for file IO with support for context manager.
+       It supports local files, Google Cloud Storage (GCS) files and
+       Google Spreadsheets.
+
+       Uses a local temporary file copy for file operations on remote files.
+       Writes to local files are written to a temporary file that is renamed
+       to the required file name on close.
+
+       This allow GCS to be used with other wrappers that use string iterables
+       such as csv.DictReader, csv.DictWriter.
+
+       To read a file:
+         with FileIO('/tmp/my-file.txt') as file:
+            for line in file:
+               print(line)
+
+       To write to a file:
+         with FileIO('tmp/my-file.txt', mode='w') as file:
+            file.write('my string')
+
+       To read a CSV file from GCS:
+         with FileIO('gs://my-bucket/some-gcs-path/gcs-file.csv', 'r') as file:
+            csv_reader = csv.DictReader(file)
+            for row in csv_reader:
+                print(row)
+
+        Note:
+        FileIO creates a local temporary copy of file opened for read/write.
```

Comment:
Is this only for GCS/spreadsheet?

Reply:
Yes, added a comment on use of temp files for reads on remote files.

It also uses temp files for writes to local files to avoid files with partial output in case the program is terminated.